### PR TITLE
ansible 2.3

### PR DESCRIPTION
**Pull request by atarkowska**
install ansible 2.3

to test use:

```
- name: openmicroscopy.devspace
  src: https://github.com/aleksandra-tarkowska/ansible-role-devspace.git
  version: ansible22
```

docker compose should work on vm without any additional upgrades

### CODE REVIEW

**Comment by joshmoore** [comment]
@aleksandra-tarkowska says there's a(n idempotency?) problem with the git module: a first run of the playbook fails, but later runs pass.

**Comment by joshmoore** [comment]
See https://github.com/openmicroscopy/ansible-role-devspace/pull/2/files#diff-2444ad0870f91f17ca6c2a5e96b26823R66

```
+# TASK [openmicroscopy.ci-devspace : Devspace stop] **********************************************************************************
 +# fatal: [a20ea3a7-7d96-4db1-aed3-c4f9de103313]: FAILED! => {"changed": false, "failed": true, "msg": "Error connecting: Error while fetching server API version: ('Connection aborted.', error(13, 'Permission denied'))"}
 +- name: Devspace stop
```

File: tasks/main.yml [hunk 1]
```diff
@@ -1,3 +1,209 @@
...
+  register: gitstatus
+
+
+# TASK [openmicroscopy.ci-devspace : Devspace stop] **********************************************************************************
+# fatal: [a20ea3a7-7d96-4db1-aed3-c4f9de103313]: FAILED! => {"changed": false, "failed": true, "msg": "Error connecting: Error while fetching server API version: ('Connection aborted.', error(13, 'Permission denied'))"}
+- name: Devspace stop
+  become: yes
+  become_user: "{{ devspace_user }}"
+  docker_service:
+    project_src: "{{ devspaceci_home }}"
```
Comment (manics) [inline tasks/main.yml]: Is `devspaceci_home` different from `devspace_home`?

File: tasks/main.yml [hunk 2]
```diff
@@ -1,3 +1,209 @@
...
+    - { srcfile: ".ssh/config", dest: ".ssh/config" }
+    - { srcfile: ".ssh/snoopycrimecop_github", dest: ".ssh/snoopycrimecop_github" }
+    - { srcfile: ".ssh/snoopycrimecop_github.pub", dest: ".ssh/snoopycrimecop_github.pub" }
+    - { srcfile: ".gitconfig", dest: ".gitconfig" }
+  when: snoopy_dir_path|length>0
+
+- name: known_hosts
+  become: yes
+  become_user: "{{ devspace_user }}"
+  shell: ssh-keyscan github.com >> {{ devspace_home }}/slave/.ssh/known_hosts
```
Comment (manics) [inline tasks/main.yml:106]: Doesn't `ssh-keyscan github.com` always give the same output? Can you hard-code it as a file instead so it's idempotent?

File: tasks/main.yml [hunk 3]
```diff
@@ -1,3 +1,209 @@
...
+  args:
+    chdir: "{{ devspace_home }}"
+    creates: "{{ devspace_home }}/nginx/sslcert"
+
+- name: rename to {{ devspace_omero_branch }}...
+  become: yes
+  become_user: "{{ devspace_user }}"
+  command: "{{ devspace_home }}/rename.py {{ devspace_omero_branch }}"
+  args:
+    chdir: "{{ devspace_home }}"
```
Comment (manics) [inline tasks/main.yml:166]: This isn't idempotent.

**Review by manics — CHANGES_REQUESTED** [review]
Changed `devspaceci_home` → `devspace_home`

Ran into this Ansible 2.3 bug: https://github.com/ansible/ansible/issues/23607 but since all docker containers are started re-running allows the remainder of the playbook to run:
`playbook.yml`:
```
- hosts: all
  roles:
    - role: openmicroscopy.docker
      docker_use_ipv4_nic_mtu: True
    - role: openmicroscopy.devspace
```
`requirements.yml`:
```
---
- src: openmicroscopy.docker
- src: openmicroscopy.versioncontrol-utils
```
`Vagrantfile`:
```
Vagrant.configure("2") do |config|
  config.vm.box = "centos/7"
  config.vm.provider "virtualbox" do |vb|
    vb.customize ["modifyvm", :id, "--memory", "2048"]
  end

  config.vm.provision "ansible" do |ansible|
    ansible.playbook = "playbook.yml"
    ansible.raw_arguments = ["--diff"]
    ansible.galaxy_role_file = "requirements.yml"
  end
end
```

**Review by manics — APPROVED** [review]

**Comment by manics** [comment]
Discussed with @aleksandra-tarkowska: the other issues were already present so shouldn't be a blocker for this PR. I've added an issue https://github.com/openmicroscopy/ansible-role-devspace/issues/3

**Comment by jburel** [comment]
This PR has been approved
I tested it again in the context of the training. It works fine
and importantly the version of docker-py installed allows to check the dynamically port assigned by https://github.com/openmicroscopy/devspace/pull/74

I have tested my change on top of this PR, and tested the full training workflow. 

If there is no objections, I would like to merge this PR (approved) and open another one on top to include the change required for the removal of the snoopy key as discussed for the training.
@joshmoore @sbesson

**Comment by joshmoore** [comment]
Based on @manics comment, no objections to merging this, but:
 * is it in a releasable state?
 * if so, what would the tag be? `0.2.0`?
 * if *not*, what other PRs are needed?

**Comment by jburel** [comment]
We will minimally need a PR to change the key name and the path before tagging i.e. ref to snoopy key and the path i.e. changes in https://github.com/openmicroscopy/ansible-role-devspace/pull/5

**Comment by jburel** [comment]
Testing set up
* this PR i.e. ansible 2.3
* devspace 0.5.2 (stable)
* Change https://github.com/openmicroscopy/infrastructure/pull/306/commits/323a40abcbbf91ae8ec4c99802746e43536e1109 

using an example playbook to provision devspace
```
---
# Exampe Playbook for provisioning Devspace

- hosts: devspace

  vars:
    # Overrides the default value in the docker role
    docker_use_ipv4_nic_mtu: True

  vars_prompt:
    - name: "devspace_jenkins_password"
      prompt: "Secure Jenkins with admin password (username: devspace)"
      private: yes

  pre_tasks:

    - fail:
        msg: "Jenkins admin password is required"
      when: devspace_jenkins_password is undefined or not devspace_jenkins_password

  roles:
    - role: openmicroscopy.devspace
      # Path to SSH and Git configuration files
      snoopy_dir_path: "/Users/jmarie/Documents/git-repo/omero-sysadmin/snoopy"
      devspace_omero_branch: develop
      devspace_git_repo: https://github.com/openmicroscopy/devspace.git
      devspace_git_repo_version: "0.5.2"
      # force a clean
      devspace_git_update: yes
      devspace_git_force: yes
```
The playbook needs to be run twice in order to deploy the devspace 
see https://10.0.51.143:8443/

**Comment by jburel** [comment]
Failure the first time the example playbook is run
```

fatal: [10.0.51.143]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "module_stderr": "Shared connection to 10.0.51.143 closed.\r\n", 
    "module_stdout": "\r\r\n\u0007Message from syslogd@devspace-training-key at Oct 30 10:43:04 ...\r\r\n kernel:unregister_netdevice: waiting for lo to become free. Usage count = 1\r\n\r\r\r\n\u0007Message from syslogd@devspace-training-key at Oct 30 10:43:14 ...\r\r\n kernel:unregister_netdevice: waiting for lo to become free. Usage count = 1\r\n\r\r\r\n\u0007Message from syslogd@devspace-training-key at Oct 30 10:43:53 ...\r\r\n kernel:unregister_netdevice: waiting for lo to become free. Usage count = 0\r\n\rTraceback (most recent call last):\r\n  File \"/tmp/ansible_pynKze/ansible_module_docker_service.py\", line 1016, in <module>\r\n    main()\r\n  File \"/tmp/ansible_pynKze/ansible_module_docker_service.py\", line 1011, in main\r\n    result = ContainerManager(client).exec_module()\r\n  File \"/tmp/ansible_pynKze/ansible_module_docker_service.py\", line 608, in exec_module\r\n    result = self.cmd_up()\r\n  File \"/tmp/ansible_pynKze/ansible_module_docker_service.py\", line 703, in cmd_up\r\n    get_stdout(fd_name)\r\n  File \"/tmp/ansible_pynKze/ansible_module_docker_service.py\", line 512, in get_stdout\r\n    new_line = re.sub(r'\\x1b\\[.+m', '', line.encode('ascii'))\r\nUnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 20: ordinal not in range(128)\r\n", 
    "msg": "MODULE FAILURE", 
    "rc": 1
}

PLAY RECAP ****************************************************************************************************************************************************************************
10.0.51.143                : ok=26   changed=4    unreachable=0    failed=1  
```

**Comment by jburel** [comment]
I think I have found an issue with this PR
I am currently doing further tests. Do not merge

**Comment by jburel** [comment]
I had to re-introduce a step https://github.com/openmicroscopy/ansible-role-devspace/pull/5/commits/8ffd104103c5128a3cebadaa2384c83634298b72
so the files have the correct owner otherwise none of the jobs will run when deploying the devspace in openstack.

**Comment by jburel** [comment]
The failure described above https://github.com/openmicroscopy/ansible-role-devspace/pull/2#issuecomment-340410752 is due to the version of ansible used
I bumped to 2.4 and there is no error

**Comment by jburel** [comment]
Closing this PR replaced by https://github.com/openmicroscopy/ansible-role-devspace/pull/5

**Comment by manics** [comment]
I've been testing ansible 2.4 with no obvious issues so far, so we could consider upgrading across the OME if you want.

**Comment by jburel** [comment]
devspace is now "self-contained"
i.e. no longer use the playbooks from infrastructure to create and provision the devspace in openstack.
https://github.com/openmicroscopy/devspace/pull/81
and it will used ansible 2.4.
As discussed in the office, I will cap it at 2.4.1 to avoid side-effect

**Comment by jburel** [comment]
saying that I started again with a "fresh" instance, The playbook runs correctly but
if I run it once, I have 
``
ERROR: No container found for nginxjenkins_1
``
I have to run it twice to get everything up....

**Pull request closed without merging**